zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs

```diff
@@ -0,0 +1,292 @@
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        param::{MAX_GAS_SIZE_IN_BYTES, NUM_ADDRESS_BYTES_USED},
+        step::ExecutionState,
+        table::{CallContextFieldTag, TxContextFieldTag},
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{
+                ConstraintBuilder, StepStateTransition,
+                Transition::{Delta, To},
+            },
+            from_bytes,
+            memory_gadget::{MemoryCopierGasGadget, MemoryExpansionGadget},
+            Cell, MemoryAddress, RandomLinearCombination,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    util::Expr,
+};
+use bus_mapping::eth_types::ToLittleEndian;
+use halo2::{arithmetic::FieldExt, circuit::Region, plonk::Error};
+use std::convert::TryInto;
+
+#[derive(Clone, Debug)]
+pub(crate) struct CallDataCopyGadget<F> {
+    same_context: SameContextGadget<F>,
+    memory_offset: MemoryAddress<F>,
+    data_offset: MemoryAddress<F>,
+    length: RandomLinearCombination<F, 4>,
+    tx_id: Cell<F>,
+    calldata_length: Cell<F>,
+    memory_expansion: MemoryExpansionGadget<F, MAX_GAS_SIZE_IN_BYTES>,
+    memory_copier_gas: MemoryCopierGasGadget<F>,
```

Comment:
I think `CallDataCopyGadget` will involve `CopyToMemoryGadget` but not add so far , right ?

Reply:
No, it won't. It only constrains next step is `CopyToMemoryGadget`.